p2p/wire_codec.go

```diff
@@ -0,0 +1,225 @@
+package p2p
+
+import (
+	"encoding/binary"
+	"errors"
+	"sync"
+)
+
+type Encoding string
+
+const (
+	Binary      Encoding = "bin"
+	Utf8        Encoding = "utf8"
+	Json        Encoding = "json"
+	Grpc        Encoding = "grpc"
+	Unsupported Encoding = "unsupported"
+)
+
+/*
+ @ Idea inspired by scuttlebutt's secure p2p wire protocol
+ @
+ @ Layout:
+ @
+ @ * = 1 byte
+ @ + = 1 bit
+ @
+ @    9 bytes header
+ @    -----header---][----body----]
+ @    [*][****][****][****....****]
+ @     0  1234  5678  ............
+ @
+ @    0: flags [ + + + + + + + + ]
+ @               <---> | | | <->
+ @      empty:____|    | | |
+ @                     | | |
+ @    body wrapped?* __| | | encoding [0,0] = binary, [0,1] = utf8, [1,1] = json
+ @                       | |
+ @          is request __| |
+ @    is erroror end ______|
+ @
+ @
+ @    *: does the body have to be decoded at the application level (i.e by the domain codec)
+ @
+ @    1234: request number as uint16, empty if not a request
+ @    5678: bodyLength
+ @
+ @ body.
+ @
+*/
+type wireCodec struct {
+	sync.RWMutex
+}
+
+func (c *wireCodec) encode(encoding Encoding, iserror bool, reqnum uint32, data []byte, wrapped bool) []byte {
+	c.Lock()
+	defer c.Unlock()
+
+	var flags byte = 0x00
+
+	bodyLength := make([]byte, 4)
```

Comment:
Can you make 4 a local constant with a semantic name

Reply:
Yes sir